### PR TITLE
feat(es/lint): implement prefer-const without options

### PR DESCRIPTION
**Pull request by ArturAralin**
**Description:**
Here is base implementations of prefer-const rule. Options I'll implement future

**BREAKING CHANGE:**

No

### CODE REVIEW

File: crates/swc_ecma_lints/src/rules/prefer_const.rs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            var_decl
+                .decls
+                .iter()
+                .for_each(|var_decl| self.collect_decl_pat(var_decl.init.is_some(), &var_decl.name))
+        }
+
+        var_decl.visit_children_with(self);
+    }
+
+    fn visit_assign_expr(&mut self, assign_expr: &AssignExpr) {
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/prefer_const.rs:216]: I think we should visit children of this node

File: crates/swc_ecma_lints/src/rules/prefer_const.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+            var_meta.postinitialized = true;
+        }
+    }
+
+    fn consider_mutation(&mut self, pat: &Pat, destructuring_assign: bool) {
+        match pat {
+            Pat::Ident(BindingIdent { id, .. }) => {
+                self.consider_mutation_for_ident(id, destructuring_assign);
+            }
+            Pat::Array(ArrayPat { elems, .. }) => elems.iter().for_each(|elem| {
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/prefer_const.rs:155]: Please use flatten instead

File: crates/swc_ecma_lints/src/rules/prefer_const.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+            }),
+            Pat::Object(ObjectPat { props, .. }) => {
+                props.iter().for_each(|prop| match prop {
+                    ObjectPatProp::KeyValue(KeyValuePatProp { value, .. }) => {
+                        self.consider_mutation(value.as_ref(), true);
+                    }
+                    ObjectPatProp::Assign(AssignPatProp { key, .. }) => {
+                        self.consider_mutation_for_ident(key, true);
+                    }
+                    _ => {}
```
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/prefer_const.rs:168]: I'm not sure, but we may need to consider rest patterns

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

File: crates/swc_ecma_lints/src/rules/prefer_const.rs (hunk 3 above)
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/prefer_const.rs:168]: Yes, you're right. eslint rule considers rest declaration

File: crates/swc_ecma_lints/src/rules/prefer_const.rs (hunk 2 above)
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/prefer_const.rs:155]: instead of `for_each`?

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/prefer_const.rs:160]: I mean `iter().flatten()`

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_lints

**Review by kdy1 — APPROVED** [review]

**Pull request merged**